### PR TITLE
[SHELL32][BROWSEUI] CDefView: Implement SFVM_ADDPROPERTYPAGES callback

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Implementing missing folder view callbacks...
JIRA issue: [CORE-19616](https://jira.reactos.org/browse/CORE-19616)

## Proposed changes

- Add `SFVM_PROPPAGE_DATA` structure to `"shobjidl.idl"`.
- Implement `CDefView::AddPropertySheetPages` by calling `SFVM_ADDPROPERTYPAGES` callback.
- Modify `CShellBrowser::DoFolderOptions` that uses `AddPropertySheetPages`.

## TODO

- [x] Do build.

### CODE REVIEW

File: dll/win32/shell32/CDefView.cpp [hunk 1]
```diff
@@ -2649,9 +2649,11 @@ HRESULT WINAPI CDefView::GetCurrentInfo(LPFOLDERSETTINGS lpfs)
 
 HRESULT WINAPI CDefView::AddPropertySheetPages(DWORD dwReserved, LPFNADDPROPSHEETPAGE lpfn, LPARAM lparam)
 {
-    FIXME("(%p) stub\n", this);
+    TRACE("(%p)->(0x%lX, %p, %p)\n", this, dwReserved, lpfn, lparam);
 
-    return E_NOTIMPL;
+    SFVM_PROPPAGE_DATA Data = { dwReserved, lpfn, lparam };
+    _DoFolderViewCB(SFVM_ADDPROPERTYPAGES, 0, (LPARAM)&Data);
+    return S_OK;
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:2656]: Should the function unconditionally return `S_OK` ? Or should it return what `_DoFolderViewCB` returns?

File: sdk/include/psdk/shobjidl.idl [hunk 2]
```diff
@@ -120,6 +120,12 @@ cpp_quote("#include <prsht.h>")
 cpp_quote("typedef LPFNADDPROPSHEETPAGE LPFNSVADDPROPSHEETPAGE;")
 cpp_quote("#endif")
 
+cpp_quote("typedef struct _SFVM_PROPPAGE_DATA {")
+cpp_quote("    DWORD                dwReserved;")
+cpp_quote("    LPFNADDPROPSHEETPAGE pfn;")
+cpp_quote("    LPARAM               lParam;")
+cpp_quote("} SFVM_PROPPAGE_DATA;")
```
Comment (HBelusca) [inline sdk/include/psdk/shobjidl.idl:127]: In official headers, this struct is _NOT_ in the idl file (and the autogenerated shobjidl.h), but is only in shlobj.h header. Please move it there instead.

**Review by HBelusca — CHANGES_REQUESTED** [review]
Besides moving the struct to the correct header, you could try revising this (currently disabled) code:
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/win32/browseui/shellbrowser.cpp;hb=4988de4b1da83978aa8669f4ee67fb06ddf927a9#l1358

File: sdk/include/psdk/shobjidl.idl (hunk 2 above)
Comment (katahiromz) [reply sdk/include/psdk/shobjidl.idl]: Moved to `<shlobj.h>`.

File: dll/win32/shell32/CDefView.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shell32/CDefView.cpp:2656]: > Should the function unconditionally return `S_OK` ?

Yes, it should.

File: sdk/include/psdk/shlobj.h [hunk 3]
```diff
@@ -110,6 +110,13 @@ typedef enum
 
 typedef int GPFIDL_FLAGS;
 
+typedef struct _SFVM_PROPPAGE_DATA
+{
+    DWORD                dwReserved;
+    LPFNADDPROPSHEETPAGE pfn;
+    LPARAM               lParam;
```
Comment (HBelusca) [inline sdk/include/psdk/shlobj.h:117]:
```suggestion
    DWORD dwReserved;
    LPFNADDPROPSHEETPAGE pfn;
    LPARAM lParam;
```

**Review by HBelusca — APPROVED** [review]

**Comment by katahiromz** [comment]
> you could try revising this (currently disabled) code

Done.

**Pull request merged**